crates/claims/crates/vc/src/syntax/non_empty_object.rs

```diff
@@ -0,0 +1,240 @@
+        self.0.push(key, value)
+    }
+
+    pub fn push_entry(&mut self, entry: Entry) -> bool {
+        self.0.push_entry(entry)
+    }
+
+    /// Push the given key-value pair to the top of the object.
+    ///
+    /// Returns `true` if the key was not already present in the object,
+    /// and `false` otherwise.
+    /// Any previous entry matching the key is **not** overridden: duplicates
+    /// are preserved, in order.
+    ///
+    /// Runs in `O(n)`.
+    pub fn push_front(&mut self, key: Key, value: Value) -> bool {
+        self.0.push_front(key, value)
+    }
+
+    pub fn push_entry_front(&mut self, entry: Entry) -> bool {
+        self.0.push_entry_front(entry)
+    }
+
+    /// Inserts the given key-value pair.
+    ///
+    /// If one or more entries are already matching the given key,
+    /// all of them are removed and returned in the resulting iterator.
+    /// Otherwise, `None` is returned.
+    pub fn insert(&mut self, key: Key, value: Value) -> Option<RemovedByInsertion> {
+        self.0.insert(key, value)
+    }
+
+    /// Inserts the given key-value pair on top of the object.
+    ///
+    /// If one or more entries are already matching the given key,
+    /// all of them are removed and returned in the resulting iterator.
+    pub fn insert_front(&mut self, key: Key, value: Value) -> RemovedByInsertFront {
+        self.0.insert_front(key, value)
+    }
+
+    /// Sort the entries by key name.
+    ///
+    /// Entries with the same key are sorted by value.
+    pub fn sort(&mut self) {
+        self.0.sort()
+    }
+
+    /// Tries to remove the entry at the given index.
+    ///
+    /// Returns an error if this would leave the object empty.
+    pub fn try_remove_at(&mut self, index: usize) -> Result<Option<Entry>, EmptyObject> {
+        if index == 0 && self.0.len() == 1 {
+            Err(EmptyObject)
+        } else {
+            Ok(self.0.remove_at(index))
+        }
+    }
+
+    /// Tries to remove all entries associated to the given key.
+    ///
+    /// Returns an error if this would leave the object empty.
+    ///
+    /// Runs in `O(n)` time (average).
+    pub fn try_remove<'q, Q>(
+        &mut self,
+        key: &'q Q,
+    ) -> Result<RemovedEntries<'_, 'q, Q>, EmptyObject>
+    where
+        Q: ?Sized + Hash + Equivalent<Key>,
+    {
+        if self.iter().all(|e| key.equivalent(&e.key)) {
+            Err(EmptyObject)
+        } else {
+            Ok(self.0.remove(key))
+        }
+    }
+
+    /// Tries to remove the unique entry associated to the given key.
+    ///
+    /// Returns an error if multiple entries match the key, or if the object
+    /// would be left empty.
+    ///
+    /// Runs in `O(n)` time (average).
+    pub fn try_remove_unique<Q>(&mut self, key: &Q) -> Result<Option<Entry>, RemoveUniqueError>
+    where
+        Q: ?Sized + Hash + Equivalent<Key>,
+    {
+        if self.iter().all(|e| key.equivalent(&e.key)) {
```

Comment:
this is confusing me -- if all keys are the same, it means it's an empty object?

Reply:
If all keys are the key you are trying to remove, then this would leave the object empty, breaking the invariant.

Reply:
Makes sense 👍